server/events/working_dir.go

```diff
@@ -89,17 +89,41 @@ func (w *FileWorkspace) Clone(
 		}
 		revParseCmd := exec.Command("git", "rev-parse", pullHead) // #nosec
 		revParseCmd.Dir = cloneDir
-		output, err := revParseCmd.CombinedOutput()
-		if err != nil {
-			log.Warn("will re-clone repo, could not determine if was at correct commit: %s: %s: %s", strings.Join(revParseCmd.Args, " "), err, string(output))
+		outputRevParseCmd, errRevParseCmd := revParseCmd.CombinedOutput()
+		if errRevParseCmd != nil {
+			log.Warn("will re-clone repo, could not determine if was at correct commit: %s: %s: %s", strings.Join(revParseCmd.Args, " "), err, string(outputRevParseCmd))
 			return w.forceClone(log, cloneDir, headRepo, p)
 		}
-		currCommit := strings.Trim(string(output), "\n")
+		currCommit := strings.Trim(string(outputRevParseCmd), "\n")
+
+		// We're bring our remote refs up to date
+		remoteUpdateCmd := exec.Command("git", "remote", "update")
+		remoteUpdateCmd.Dir = cloneDir
+		outputRemoteUpdate, errRemoteUpdate := remoteUpdateCmd.CombinedOutput()
+		if errRemoteUpdate != nil {
+			log.Warn("getting remote update failed: %s", string(outputRemoteUpdate))
+		}
+
+		// We're checking if remote master branch has diverged
+		statusUnoCmd := exec.Command("git", "status", "-uno")
+		statusUnoCmd.Dir = cloneDir
+		outputStatusUno, errStatusUno := statusUnoCmd.CombinedOutput()
+		if errStatusUno != nil {
+			log.Warn("getting repo status has failed: %s", string(outputStatusUno))
+		}
+		status := strings.Trim(string(outputStatusUno), "\n")
+		hasDiverged := strings.Contains(status, "have diverged")
+		if hasDiverged {
```

Comment:
Shouldn't we skip this section if any of the above commands fail? Maybe pull it into its own function so you can return early.

Reply:
I did it like this because I don't want to impact the clone function if for any reason it cannot detected that master branch has diverged as it displays only a warning in the comment of the plan.